cumulusci/utils/yaml/cumulusci_yml.py

```diff
@@ -1,4 +1,276 @@
-from .safer_loader import load_yaml_data
+import re
+from logging import getLogger
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Union, Sequence
 
-# this will be replaced with a Pydantic-based loader soon
-cci_safe_load = load_yaml_data
+from typing_extensions import Literal, TypedDict
+from pydantic import Field
+
+from cumulusci.utils.fileutils import DataInput, load_from_source
+from cumulusci.utils.yaml.model_parser import CCIDictModel
+from cumulusci.utils.yaml.safer_loader import load_yaml_data
+
+default_logger = getLogger(__name__)
+
+
+NBSP = "\u00A0"
+
+pattern = re.compile(r"^\s*[\u00A0]+\s*", re.MULTILINE)
+
+
+#  type aliases
+PythonClassPath = str
+URL = str
+
+
+class FlowReference(CCIDictModel):
+    flow: str = None
+    options: Dict[str, Any] = {}
+    ignore_failure: bool = None  # is this allowed?
+    when: str = None  # is this allowed?
+    ui_options: Dict[str, Any] = {}
+
+
+class TaskReference(CCIDictModel):
+    task: str
+    options: Dict[str, Any] = {}
+    ui_options: Dict[str, Any] = {}
+    ignore_failure: bool = None
+    when: str = None  # is this a documented feature?
+    # classpath, description allowed?
+
+
+class Task(CCIDictModel):
+    name: str = None  # is this real or a typo
+    class_path: str = None
+    description: str = None
+    options: Dict[str, Any] = None
+    group: str = None
+    ui_options: Dict[str, Any] = None
+
+
+class Flow(CCIDictModel):
+    description: str = None
+    steps: Dict[str, Union[FlowReference, TaskReference]]
+    group: str = None
+
+
+class Package(CCIDictModel):
+    name: Optional[str] = None
+    name_managed: Optional[str] = None
+    namespace: Optional[str] = None
+    install_class: str = None
+    uninstall_class: str = None
+    api_version: str = None
+
+
+class Test(CCIDictModel):
+    name_match: str
+
+
+class Parser(CCIDictModel):
+    class_path: PythonClassPath
+    title: str
+
+
+class ReleaseNotes(CCIDictModel):
+    parsers: Dict[int, Parser]  # should probably be a list
+
+
+class Git(CCIDictModel):
+    repo_url: str = None
+    default_branch: str = None
+    prefix_feature: str = None
+    prefix_beta: str = None
+    prefix_release: str = None
+    push_prefix_sandbox: str = None
+    push_prefix_production: str = None
+    release_notes: ReleaseNotes = None
+    two_gp_context: str = Field(None, alias="2gp_context")
+
+
+class ApexDoc(CCIDictModel):
+    homepage: str = None  # can't find these in the CCI Python code
+    banner: str = None
+    branch: str = None
+    repo_dir: Path = None
+
+
+class Check(CCIDictModel):
+    when: str = None
+    action: str = None
+    message: str = None
+
+
+class Plan(CCIDictModel):  # MetaDeploy plans
+    title: str
+    description: str = None
+    tier: str = None
```

Comment:
enum choice between primary, secondary, additional

Reply:
If I use enums, consuming code will need to make more pervasive changes when it starts consuming the models instead of the YAML parser output.

Reply:
Isn't that the goal -- to be able to make the more pervasive changes enabled by being able to assume the input is already validated?

Reply:
My goal is to give better error messages by catching errors earlier in the project lifespan (of an individual CCI project) and earlier in a process execution process. Rewriting "string code" to "enum code" doesn't do a lot to advance those. I can do it if you insist but it just seems like a different project to figure out how and when we consume these models.

Reply:
The reason I suggested an enum was because I figured then the error can be more helpful by indicating what the valid options are -- which I think is in service of the overall goal here. That can probably be done either using an enum or by a custom validator that still returns a string. And I see this as an opportunity but not a blocker for this PR.

Reply:
Okay yeah, I'll do this instead:

```
    tier: Literal["primary", "secondary", "additional"] = None
```

Which generates:

```
'foo.yml' -> 'plans' -> 'plan1' -> 'tier' : unexpected value; permitted: 'primary', 'secondary', 'additional'
```